### PR TITLE
Serializable lock

### PR DESCRIPTION
**Pull request by mrocklin**
And use in `dask.array.from_array` and for eventual use with distributed xarray.

cc @shoyer @pwolfram

### CODE REVIEW

**Comment by mrocklin** [comment]
@shoyer how does XArray handle locking many from_array calls with the same lock?  Do you create an external lock for HDF5 and then apply that same lock each time?

**Comment by shoyer** [comment]
>  how does XArray handle locking many from_array calls with the same lock? Do you create an external lock for HDF5 and then apply that same lock each time?

Indeed, we default to a single global lock for all HDF5 read/write (currently just called `_global_lock`):
https://github.com/pydata/xarray/blob/2a17af77955a11e93fafccdc8bc619adfc9e402a/xarray/backends/api.py#L50

**Comment by mrocklin** [comment]
So, if we use this PR then you would change 

```python
_global_lock = threading.Lock()
```

to 

```python
from dask.utils import SerializableLock
_global_lock = SerializableLock()
```

or 

```python
from dask.utils import SerializableLock
_global_lock = SerializableLock('hdf5')  # if we want to have a consistent lock across projects
```

File: dask/utils.py [hunk 1]
```diff
@@ -897,3 +899,64 @@ class MethodCache(object):
...
+
+    This is useful for consistently protecting resources on a per-process
+    level.
+
+    The creation of locks is itself not threadsafe.
+    """
+    def __init__(self, token=None):
+        self.lock = Lock()
+        self.token = token or str(uuid.uuid4())
+        SerializableLock._locks[self.token] = self.lock
```
Comment (pitrou) [inline dask/utils.py]: I don't know if serializable locks are expected to be created in large numbers, but you may want to use a `WeakValueDictionary` here if you don't want locks to be immortal.

**Comment by shoyer** [comment]
Yes, something like `SerializableLock('hdf5')` seems ideal.

File: dask/utils.py (hunk 1 above)
Comment (shoyer) [inline dask/utils.py]: What happens if `token` already exists in `_locks`? e.g., two different libraries create a `SerializableLock('hdf5')`?

Comment (mrocklin) [inline dask/utils.py]: Nice.  I was unfamiliar with `WeakValueDictionary`.  Thanks!

File: dask/utils.py [hunk 2]
```diff
@@ -897,3 +900,67 @@ class MethodCache(object):
...
+
+    @property
+    def locked(self):
+        return self.locked
+
+    def __getstate__(self):
+        return self.token
+
+    def __setstate__(self, token):
+        self.token = token
```
Comment (shoyer) [inline dask/utils.py]: This is the exact same logic as in `__init__` now -- consolidate them?

Comment (mrocklin) [inline dask/utils.py]: Fixed

**Comment by mrocklin** [comment]
Merging soon if no further comments.  @shoyer does this suffice?

File: dask/utils.py [hunk 3]
```diff
@@ -897,3 +900,62 @@ class MethodCache(object):
...
+    """
+    def __init__(self, token=None):
+        self.token = token or str(uuid.uuid4())
+        if self.token in SerializableLock._locks:
+            self.lock = SerializableLock._locks[self.token]
+        else:
+            self.lock = Lock()
+            SerializableLock._locks[self.token] = self.lock
+
+    def acquire(self, *args):
```
Comment (shoyer) [inline dask/utils.py]: Lets either switch this to a normal argument `blocking=True` or at least allow passing on a keyword argument as well. `lock.acquire(blocking=True)` is much more readable.

File: dask/tests/test_utils.py [hunk 4]
```diff
@@ -213,3 +214,49 @@ def test_skip_doctest():
...
+        assert not a.acquire(False)
+
+    a2 = pickle.loads(pickle.dumps(a))
+    a3 = pickle.loads(pickle.dumps(a))
+    a4 = pickle.loads(pickle.dumps(a2))
+
+    for x in [a, a2, a3, a4]:
+        for y in [a, a2, a3, a4]:
+            with x:
+                assert not y.acquire(False)
```
Comment (shoyer) [inline dask/tests/test_utils.py]: as mentioned below, switch this to `y.acquire(blocking=False)`

File: dask/utils.py (hunk 3 above)
Comment (mrocklin) [inline dask/utils.py]: I ended up reverting this for py27 compatibility but you're right that we can be py3 style compatible up front and defend py27 when calling acquire ourselves.

Comment (shoyer) [inline dask/utils.py]: It doesn't work to write `lock.acquire(blocking=True)` on Python 2.7?

**Comment by mrocklin** [comment]
In [1]: from threading import Lock

In [2]: lock = Lock()

In [3]: lock.acquire(blocking=True)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-74ce909888d4> in <module>()
----> 1 lock.acquire(blocking=True)

TypeError: acquire() takes no keyword arguments

**Comment by shoyer** [comment]
wtf python 2!

fine, let's merge as is.

**Pull request merged**